### ISSUE TITLE
Undoing a replace behaves incorrectly when the collab module is active

### ISSUE THREAD

**Issue by kiejo** [ProseMirror/prosemirror#292]
I tested this using the latest master (rewritten history module) and could only reproduce the error when using the `collab` module, e.g.:
```javascript
require("../dist/collab")
var pm = new ProseMirror({
  /* options */
  collab: {version: 0}
})
```
Steps to reproduce:
1. Type `hello world`
2. Select `world`
3. Type `test` (-> you now have `hello test` in the editor)
4. Press undo multiple times
Problem: The word `world` stays there and does not get removed.

Doing the same thing in an editor without the `collab` module, works as expected.

**Comment by kiejo** [comment]
I found another misbehavior that's probably related to this as it only occurs when the `collab`module is active.  
Steps to reproduce:
1. Create a bullet list with a nested child like this:
![bullet_undo](https://cloud.githubusercontent.com/assets/152507/14389806/91fd7124-fdb4-11e5-9f17-74b17f106536.png)
2. Press undo multiple times
Problem: The text is removed, but the bullets stay:
![bullet_undo2](https://cloud.githubusercontent.com/assets/152507/14389846/bb03d482-fdb4-11e5-9a41-1b9cd42d4b52.png)

**Comment by marijnh** [comment]
I followed these steps (adjusted demo to start with empty document with collab module active), but everything is working as expected -- no errors, undo and redo go through exactly as hoped.

Any other factor that might be relevant in your setup?

**Comment by kiejo** [comment]
That's strange as I did the exact same thing. Here's the code I have in `demo.js`:
```javascript
var ProseMirror = require("../dist/edit/main").ProseMirror
require("../dist/inputrules/autoinput")
require("../dist/menu/tooltipmenu")
require("../dist/menu/menubar")
require("../dist/collab")

var pm = window.pm = new ProseMirror({
  place: document.querySelector(".full"),
  autoInput: true,
  tooltipMenu: {selectedBlockMenu: true},
  menuBar: {float: true},
  doc: document.querySelector("#content"),
  docFormat: "dom",
  collab: {version: 0}
})

document.querySelector("#mark").addEventListener("mousedown", function(e) {
  pm.markRange(pm.selection.from, pm.selection.to, {className: "marked"})
  e.preventDefault()
})
```
I tested this in Chrome on Linux and ran the demo using `npm run dist-watch` and `npm run demo`.
Are you sure you're calling `require("../dist/collab")`? Removing this line does not cause any errors (even with the collab option specified), but prevents the undo misbehavior.

**Comment by marijnh** [comment]
Yes, I definitely have the collab module loaded. (And am also on Chrome Linux).

Since you didn't delete the `doc` option, you are apparently starting with a non-empty document. Could you list the _precise_ steps you're taking to get the error (do you clear the document first? create a new paragraph? well, I tried those two, but maybe you're doing something different)

**Comment by kiejo** [comment]
I now reduced the `demo.js` file to the following and can still reproduce the undo problem:
```javascript
var ProseMirror = require("../dist/edit/main").ProseMirror
require("../dist/collab")

var pm = window.pm = new ProseMirror({
  place: document.querySelector(".full"),
  collab: {version: 0}
})
```

Just to be sure, I also cloned a fresh copy of the repository and ran `npm install` followed by `npm run demo`.
Here's a recording of what I did. After finishing typing `test`, I pressed `Ctrl+Z` multiple times. I hope this helps. I'm not sure what else could be related to this issue. No errors are thrown in the console, it's just that the undo misbehaves. As soon as I remove the `collab` module, it works fine.
![out](https://cloud.githubusercontent.com/assets/152507/14403465/6e0179b6-fe57-11e5-8200-592b8b07d504.gif)
